snow/engine/snowman/engine_decorator.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package snowman
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+)
+
+type decoratedEngineWithStragglerDetector struct {
+	*Engine
+	sd *stragglerDetector
+	f  func(time.Duration)
+}
+
+func NewDecoratedEngineWithStragglerDetector(e *Engine, time func() time.Time, f func(time.Duration)) common.Engine {
+	minConfRatio := float64(e.Params.AlphaConfidence) / float64(e.Params.K)
+	sd := newStragglerDetector(time, e.Config.Ctx.Log, minConfRatio, e.Consensus.LastAccepted,
+		e.Config.ConnectedValidators.ConnectedValidators, e.Config.ConnectedValidators.ConnectedPercent,
+		e.Consensus.Processing, e.acceptedFrontiers.LastAccepted)
+	return &decoratedEngineWithStragglerDetector{
+		Engine: e,
+		f:      f,
+		sd:     sd,
+	}
+}
+
+func (de *decoratedEngineWithStragglerDetector) Chits(ctx context.Context, nodeID ids.NodeID, requestID uint32, preferredID ids.ID, preferredIDAtHeight ids.ID, acceptedID ids.ID) error {
+	behindDuration := de.sd.CheckIfWeAreStragglingBehind()
+	if behindDuration > 0 {
+		de.Engine.Config.Ctx.Log.Info("We are behind the rest of the network", zap.Float64("seconds", behindDuration.Seconds()))
+	}
+	de.Engine.metrics.stragglingDuration.Set(float64(behindDuration))
+	de.f(behindDuration)
```

Comment:
This is not for testing. This is a decorator, so it invokes the method it decorates but also invokes a function, which is pluggable. The pluggable function it invokes, if this `f`.

Reply:
I still slightly think this is a bit extra (since we don't actually use it in production). but if you think we should keep it, can we rename it and add comment so that it's visible for consumers?